cdc/http_handler.go

```diff
@@ -260,6 +280,100 @@ func (s *Server) handleChangefeedQuery(w http.ResponseWriter, req *http.Request)
 	writeData(w, resp)
 }
 
+// handleChangefeeds dispatch the request to the specified handleFunc according to the request method.
+func (s *Server) handleChangefeeds(w http.ResponseWriter, req *http.Request) {
+	if req.Method == http.MethodGet {
+		s.handleChangefeedsList(w, req)
+		return
+	}
+}
+
+func (s *Server) handleChangefeedsList(w http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodGet {
+		writeError(w, http.StatusBadRequest, cerror.ErrSupportPostOnly.GenWithStackByArgs())
+		return
+	}
+	s.ownerLock.RLock()
+	defer s.ownerLock.RUnlock()
+	if s.owner == nil {
+		handleOwnerRespJSON(w, concurrency.ErrElectionNotLeader)
+		return
+	}
+
+	err := req.ParseForm()
+	if err != nil {
+		writeInternalServerErrorJSON(w, cerror.WrapError(cerror.ErrInternalServerError, err))
+		return
+	}
+
+	statuses, err := s.owner.etcdClient.GetAllChangeFeedStatus(req.Context())
+	changefeedIDs := make(map[string]struct{}, len(statuses))
+	if err != nil {
+		writeInternalServerErrorJSON(w, err)
+		return
+	}
+	for cid := range statuses {
+		changefeedIDs[cid] = struct{}{}
+	}
+
+	state := req.Form.Get(APIOpVarChangefeeds)
+	filters := map[model.FeedState]struct{}{
+		model.StateNormal:   {},
+		model.StateStopped:  {},
+		model.StateFailed:   {},
+		model.StateFinished: {},
+		model.StateRemoved:  {},
+		model.StateError:    {},
+	}
+	switch state {
+	case "all":
+		filters = make(map[model.FeedState]struct{})
+	case "":
+		delete(filters, model.StateNormal)
+		delete(filters, model.StateStopped)
+		delete(filters, model.StateFailed)
+	default:
+		delete(filters, model.FeedState(state))
+	}
+
+	resps := make([]*ChangefeedCommonInfo, 0)
+	for changefeedID := range changefeedIDs {
+		cf, status, feedState, err := s.owner.collectChangefeedInfo(req.Context(), changefeedID)
+		if err != nil && cerror.ErrChangeFeedNotExists.NotEqual(err) {
+			writeInternalServerErrorJSON(w, err)
+			return
+		}
+		if _, ok := filters[feedState]; ok {
```

Comment:
could u extract filter as a func, maybe do not need to build the filter map per request ?

Reply:
done